### ISSUE TITLE
Some chat issues in production server

### ISSUE THREAD

**Issue by Silver-IT** [okTurtles/group-income#1507]
### Problem

1. It makes error when try to create direct message.
2. In `General` chatroom, we should see the `Joined General` message (which is created automatically when a new member is joined the group). But we can not see them.

@taoeffect, described issues in detail here in Slack.
https://okturtles.slack.com/archives/C0EH7P20Y/p1677716969380129

### Solution

I am not sure what the reason is at this moment. Need to debug.

**Comment by Silver-IT** [comment]
@taoeffect, I think I know the reason why it fails to create DM in the live server.
The mailbox contract is created with the version of 0.0.1 (or anyone) which is earlier version.
The earlier version creates the mailbox in this format. 
```
'gi.contracts/mailbox': {
  validate: object, // TODO: define this
  process ({ data }, { state }) {
    for (const key in data) {
      Vue.set(state, key, data[key])
    }
    Vue.set(state, 'messages', [])
  }
}
```
But in the new version of mailbox contract, it creates in the following format.
```
'gi.contracts/mailbox': {
  validate: objectOf({ username: string }),
  process ({ data }, { state }) {
    const initialState = merge({
      attributes: {
        creator: data.username,
        autoJoinAllowance: true
      },
      chatRooms: {}
    }, data)
    for (const key in initialState) {
      Vue.set(state, key, initialState[key])
    }
  }
}
```
And in `createDirectMessage` mailbox contract function, we are calling `state.attributes.creator` but `state.attributes` doesn't exist because this mailbox contract is created with the earlier version.

I think we should create an issue for this. **Updating the state format in the new version can cause error in contract function call**.

**Comment by taoeffect** [comment]
@Silver-IT Looking at the code above, I just realized that I missed something during my review: the validation function of a contract should almost always be `objectMaybeOf`, not `objectOf`, as that helps with migrating contracts using the data of old contracts.

So this:

```js
validate: objectOf({ username: string }),
```

Should be:

```js
validate: objectMaybeOf({ username: string }),
```

*Although it should really be `object` (see other comments below about using `meta.username` instead of `data.username`)

This makes it possible to create a new contract using the data of an old contract without any validation errors being thrown (which is again, might be important for when we implement migrating from server to server).

> I think we should create an issue for this. Updating the state format in the new version can cause error in contract function call.

There is no generic way to fix this. The correct way is to remember and understand how the contracts were written in previous versions, and either (a) adjust your code accordingly (for example, by using `vueFetchInitKV`), (b) to re-create and migrate the contracts, (c) add a hack to the code to work around this bug (temporarily), or (d) start over by recreating the group.

In this case, option (a) is not available to you, because you do not know what the default value of `creator` should be. Option (b) is possible, but complicated. That leaves (c) and (d). We just re-created the group, so let's maybe try (c)?

The hack I would suggest would involve modifying `'state/vuex/postUpgradeVerification'` to contain an event listener for the `LOGIN` event, which is emitted at the very end of the login procedure. Something like this:

```js
sbp('okTurtles.events/on', LOGIN, ({ username, identityContractID }) => {
  const getters = sbp('state/vuex/getters')
  if (!getters.currentMailboxState.attributes) {
    // TODO: send a 'gi.contracts/mailbox/setAttributes' message
    //       to set the creator and autoJoinAllowance properties
  }
})
```

Then, you will also need to modify `'gi.contracts/mailbox/setAttributes'` validation to be more lenient (check for `state.attributes` existing):

```js
validate: (data, { state, meta }) => {
  if (state.attributes && state.attributes.creator !== meta.username) {
    throw new TypeError(L('Only the mailbox creator can set attributes.'))
  }
  object(data)
},
```

Also, I'm surprised that I didn't point this out in my review as well, but I don't think we should be using `data.username` to set the `creator` attribute in the contract's constructor. Instead, please use `meta.username` to avoid unnecessarily duplicating that info. The first person to create a mailbox should be the creator, and the username is already included as part of the metadata. Can you update it to do this?

**Comment by Silver-IT** [comment]
@taoeffect, thanks for your detailed description. Actually, I noticed that the only solution depends on us when we update the contract. We need to take care and to consider the older versions so that it wouldn't break.

And I'd like to say that `mailbox` contract is created before `identity` contract is created. So I don't think it's possible to use `meta.username` instead of `data.username` in the `gi.contracts/mailbox` mailbox contract constructor.

**Comment by taoeffect** [comment]
@Silver-IT 

> So I don't think it's possible to use meta.username instead of data.username in the gi.contracts/mailbox mailbox contract function.

It is possible. You can modify the contract right now to do it. Versions from that point forward, when created, will work correctly.

**Comment by Silver-IT** [comment]
@taoeffect, I could be a bit picky, but I would like to completely make it confirmed.

1. Why almost of contract validation functions should be `'objectMaybeOf'` instead of `'objectOf'`? I think the **same version of validation function** would be chosen and executed for a specific version of event. So I think we should be careful **only in using the state** which could be made using the older version of contract events.
2. For the solution to fix the current version of mailbox contract, you recommended to listen `LOGIN` event inside `'state/vuex/postUpgradeVerification'` function and make a `'gi.contracts/mailbox/setAttributes'` event if some mandatory attributes are not set. So far, I thought `'state/vuex/postUpgradeVerification'` function is for fixing the mismatch of the global state, not the contract state. Did I think right? And why can't we make that event at the end of `'gi.actions/identity/login'` function? It could be more simple, I think.
3. Now, we create mailbox contract before creating identity contract passing the mailbox contract address as input data. And we set the `meta.username` of mailbox contract from the `sbp('state/vuex/state').loggedIn`, so we can not use the `meta.username` in mailbox contract constructor at the moment. In order to use `meta.username` in mailbox contract constructor, we need **to update the workflow of signup and login** or **to update the function to create mailbox metadata getting the temporarily saved username by calling** `sbp('okTurtles.data/get')`. Which way do you recommend? a) Keep using `data.username` b) Update the function to create mailbox contract metadata c) Update Register, Login workflow and identity contract. To be honest, I recommend a), but I don't think you recommend it.

**Comment by Silver-IT** [comment]
> In General chatroom, we should see the Joined General message (which is created automatically when a new member is joined the group). But we can not see them.

This is the second problem described in this ticket. I think this is because of changes on the chatroom contract.
I updated the variable name `saveMessage` to `onlyRenderMessage` for better understanding but this change breaks the previous messages being rendered. That `onlyRenderMessage` is being used for render messages only in the Chat page and it is set in the `ChatMain.vue` file.

To fix these two issues described in this ticket, there could be many hacks but I think the best solution is to reset the database and restart the server.
I will be really careful in updating the contract from now.

**Comment by taoeffect** [comment]
> 1. Why almost of contract validation functions should be 'objectMaybeOf' instead of 'objectOf'? I think the same version of validation function would be chosen and executed for a specific version of event. So I think we should be careful only in using the state which could be made using the older version of contract events.

Say you want to move from Server 1 to Server 2. Server 1 has version A of the mailbox contract, and Server 2 has version B. If you use `objectOf`, there is a high probability that you won't be able to migrate the contract data. On the other hand, if you use `objectMaybeOf`, it is more likely that you will be able to copy the entire state. This is why constructors must use `objectMaybeOf`.

> And why can't we make that event at the end of 'gi.actions/identity/login' function? It could be more simple, I think.

It would not be more simple - it would complicate the `'gi.actions/identity/login'` selector with unnecessary hacks.

This bug exists because the contracts were not properly upgraded to begin with. `'gi.actions/identity/login'` **must not care** about the contract-specific changes that we're making. This is why hacks for upgrading contracts go into `'state/vuex/postUpgradeVerification'`.

> Which way do you recommend? a) Keep using data.username

I agree, (a) is better! I forgot about that!

> To fix these two issues described in this ticket, there could be many hacks but I think the best solution is to reset the database and restart the server.

Yeah I can do that. Still, the mailbox contact needs to use `objectMaybeOf` for its constructor - don't forget to make that change in your next PR.

>  I will be really careful in updating the contract from now.

👍

**Comment by taoeffect** [comment]
> To fix these two issues described in this ticket, there could be many hacks but I think the best solution is to reset the database and restart the server.

Actually, I don't understand. I think there's only 1 hack to fix this, the one I described above to `setAttributes` within a `LOGIN` event listener inside of `postUpgradeVerification`. What else is there?

**Comment by Silver-IT** [comment]
> Actually, I don't understand. I think there's only 1 hack to fix this, the one I described above to setAttributes within a LOGIN event listener inside of postUpgradeVerification. What else is there?

I agree we can fix the first issue by adding event listener in `postUpgradeVerifiation`, but we can not fix the second issue.
I think there is no way except `to reset the database`. It's actually a big mistake which I made in the last PR.

**Comment by Silver-IT** [comment]
1. I totally understand why we need to use `objectMaybeOf` in the constructor. It's my misunderstanding but I thought we should use `objectMaybeOf` in all the contract functions.
2. I agree that we should keep the **Coding Convention** and add the hack in `postUpgradeVerification` if needed.

As a result, I will update the mailbox constructor validation function and make it ready to be reviewed, so that we can merge this and reset the database and make the live server working.

**Comment by taoeffect** [comment]
> I agree we can fix the first issue by adding event listener in postUpgradeVerifiation, but we can not fix the second issue.

Sorry, which is the second issue?

**Comment by Silver-IT** [comment]
> Sorry, which is the second issue?

In General chatroom, we should see the Joined General message (which is created automatically when a new member is joined the group). But we can not see them.

**Comment by taoeffect** [comment]
Oh, OK, yeah, that's OK. It's not a big deal.

**Comment by Silver-IT** [comment]
Then, let's merge the PR #1508 and reset the database.

**Comment by taoeffect** [comment]
Why do we need to reset the database?

**Comment by Silver-IT** [comment]
If not, the previous message won't be visible. That's because the previous chatroom contract renders message according to the attribute `saveMessage` while the new chatroom contract renders messages according to the `onlyRenderMessage` attribute.
And this attribute is only set and used in `ChatMain.vue` which is regardless to the contract version.

**Comment by taoeffect** [comment]
But we just agreed that it's OK for the earlier messages to not be visible. It's just messages saying people joined the chatroom. It's not important enough to reset the database for.

**Comment by Silver-IT** [comment]
I thought `it` refered to `to reset database`. I didn't add event listener in the `postUpgradeVerification` because it's temporary hack and we don't need it after we reset the database. I only updated the mailbox contract constructor validate function using `objectMaybeOf` in the PR #1508, so that it could be fine to the future migrating.

I think to reset the database will fix the two issues described in this ticket and this ticket gives me yellow card in updating contracts.

**Comment by taoeffect** [comment]
Looks like problems (1) and (2) are still problems :-\

**Comment by Silver-IT** [comment]
@taoeffect, correct the two problems are still there in the live server.
1. First issue
- In `postUpgradeVerification`, we listen `LOGIN` event and update the mailbox state if it doesn't have `attributes` key.
Since `LOGIN` event is emitted at the end of login process, mailbox contract is synced in advance. While syncing the `mailbox` contract, `createDirectMessage` and `joinDirectMessage` will make errors because `attributes` key doesn't exist in the state yet.
- We make `attributes` inside the state when the user is logged in. So `attributes` key can't be there unless user logs in at least once.
- I did focus only on the `attributes`. But there is `chatRooms` which is also newly added to mailbox state in the new version. 

**I couldn't guess all these before. That's my mistake.**

2. Second issue
- As I mentioned [here](https://github.com/okTurtles/group-income/issues/1507#issuecomment-1457286583), it's not possible to render messages which are created before unless we reset the database. Only newly created messages can be visible.

I tried to find a solution (of course, it is technically possible) and I think it would be a bit messy because we should consider the three things. Sorry to ask again, but we should reset the database. I think that's the smart solution at the moment.
